code-studio/src/js/pd/workshop_dashboard/attendance/workshop_attendance.jsx

```diff
@@ -0,0 +1,194 @@
+  contextTypes: {
+    router: React.PropTypes.object.isRequired
+  },
+
+  propTypes: {
+    params: React.PropTypes.shape({
+      workshopId: React.PropTypes.string.isRequired,
+      sessionIndex: React.PropTypes.string
+    }).isRequired
+  },
+
+  getInitialState: function () {
+    return {
+      loading: true
+    };
+  },
+
+  componentDidMount: function () {
+    // Response format:
+    // [
+    //   state: _workshop state_,
+    //   session: {id. start, end},
+    //   attendances: [{name, email, enrolled, user_id, in_section, attended}]
+    // ]
+    this.loadRequest = $.ajax({
+      method: "GET",
+      url: `/api/v1/pd/workshops/${this.props.params.workshopId}/attendance`,
+      dataType: "json"
+    }).done(function (data) {
+      this.setState({
+        loading: false,
+        workshopState: data.state,
+        sessionAttendances: data.session_attendances,
+        adminActions: data.admin_actions
+      });
+    }.bind(this));
+  },
+
+  componentWillUnmount: function () {
+    if (this.loadRequest) {
+      this.loadRequest.abort();
+    }
+    if (this.saveRequest) {
+      this.saveRequest.abort();
+    }
+  },
+
+  handleNavSelect: function (sessionIndex) {
+    this.context.router.replace(`/workshops/${this.props.params.workshopId}/attendance/${sessionIndex}`);
+    this.setState(this.state);
+  },
+
+  handleCancelClick: function (e) {
+    this.context.router.push(`/workshops/${this.props.params.workshopId}`);
+  },
+
+  handleSaveClick: function (e) {
+    var url = `/api/v1/pd/workshops/${this.props.params.workshopId}/attendance`;
+    var data = this.prepareDataForApi();
+    this.saveRequest = $.ajax({
+      method: 'PATCH',
+      url: url ,
+      dataType: 'json',
+      contentType: 'application/json',
+      data: JSON.stringify({pd_workshop: data})
+    }).done(function () {
+      this.context.router.push('/workshops/' + this.props.params.workshopId);
+    }.bind(this));
+  },
+
+  prepareDataForApi: function () {
+    // Convert to {session_attendances: [session_id, attendances: [user_id or email]]}
+    return {
+      session_attendances: this.state.sessionAttendances.map(function (sessionAttendance) {
+        return {
+          session_id: sessionAttendance.session.id,
+          attendances: sessionAttendance.attendance.filter(function (attendance) {
+            return attendance.attended;
+          }).map(function (attendance) {
+            if (attendance.user_id) {
+              return {id: attendance.user_id};
+            }
+            // For admin-override mode, the user may not have an account so use email.
+            // In that case, an account will be created in the backend, invited by the admin.
+            return {email: attendance.email};
+          })
+        };
+      })
+    };
+  },
+
+  handleAttendanceChange: function (i, value) {
+    this.state.sessionAttendances[this.activeSessionIndex()].attendance[i].attended = value;
```

Comment:
I like using [Immutable.js](https://facebook.github.io/immutable-js/) for this reason. It provides nice APIs for cloning deep state with deep modications.